### ISSUE TITLE
Contract State Machine

### ISSUE THREAD

**Issue by soupglasses** [cs-24-sw-3-15/contract-service#42]
Contracts should get uploaded into a state of Pending.

This then gets approved by an Admin or Legal Councellor. It now becomes Approved.

```rb
  status = [ :pending, :approved, :denied ] # enum in db/migrations
  lifecycle = [ :inactive, :active, :replaced, :expired ] # enum in db/migrations
```

Here is some rough code from me quickly verifying that the implementation can work using the state_machines gem. 

```rb
# app/models/contract.rb
  state_machine :initial => :pending do
    # Should an active contract still be active if it gets replaced?
    # - Do we care about "active but replaced" contracts, or are they simply replaced?
    # - If a replacement gets filled in, automatically remove from Active state?
    # when admin approves :approved
    # when :approved and start_date is hit, it becomes :active
    # when admin denies :pending it becomes :denied, done
    # if its :active it can after expiry date become :expired, done
    # it can always be :replaced??
    #:approved
    #:denied
  end

  def active?
    # state == :active and start_date > DateTime.now
    # SHOULD ONLY WORK WHEN start_date IS PASSED AND end_date IS NOT PASSED.
  end
```

Focus on getting the two state machines implemented, likely using aasm, as it got support for multiple state machines per model.
https://github.com/aasm/aasm?tab=readme-ov-file#multiple-state-machines-per-class

**Comment by soupglasses** [comment]
Bonus is looking into ActiveJob for scheduling the activation and de-activation stages of a contract. But this can be pulled out into a separate issue.

**Comment by soupglasses** [comment]
For storing it into activerecord https://github.com/aasm/aasm?tab=readme-ov-file#activerecord

**Comment by soupglasses** [comment]
Replacement is now extracted into a separate issue #44 to focus on the state machine first.

**Comment by soupglasses** [comment]
Test case ideas. Distribute out into multiple tests as needed.

```rb
contract = Contract.new

contract.start_date = 1.minute.ago
contract.save
# Ensure state is :active

contract.start_date = 1.minute.from_now
contract.save
# Ensure state has now been changed to :queued
# Ensure ActiveJob gets queued.

# Change time forwards x minutes using test framework. Good idea to look up ActiveJob Testing.
# Ensure state has now been changed to :active

# Test :expired to :active through changing start_date
# Test :expired to :queued through changing start_date
```

**Comment by soupglasses** [comment]
Closed by #45
